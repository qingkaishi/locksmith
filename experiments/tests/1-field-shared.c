#include <pthread.h>

struct pair {
  int data;  //this should be shared, and report a race
  int data2; //this shouldn't be shared nor exist in the .dot file
};

void *f(struct pair *p) {
  p->data++;
  return NULL;
}

int main() {
  pthread_t t1, t2;
  struct pair *p = (struct pair*) malloc(sizeof (struct pair));

  pthread_create(&t1, NULL, f, p);
  pthread_create(&t2, NULL, f, p);
  return 1;
}
