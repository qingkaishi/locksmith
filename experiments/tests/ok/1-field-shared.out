************************* STARTING *************************

shared: alloc.data:1-field-shared.c:15

Warning: Possible data race: alloc.data:1-field-shared.c:15 is not protected!
 references:
  dereference of *p.data:1-field-shared.c:8 at 1-field-shared.c:9
    alloc.data:1-field-shared.c:15 => p->data:1-field-shared.c:15
                                   => *p.data:1-field-shared.c:8 at f 1-field-shared.c:17
  locks acquired:
    <empty>
  in: main at 1-field-shared.c:13 


*************************** DONE ***************************
